File: turbo/apps/desktop/src/desktop-auth-session.ts
import { authContract } from "@okouai/api-contracts/contracts/auth";
import type { DesktopAuthWindowRequest } from "./desktop-auth-window";
import type { DesktopAuthState } from "./desktop-bridge";
import type { DesktopAuthCallback } from "./desktop-auth";
import type { DesktopClientHeaderInjector } from "./desktop-client-headers";
import { singleFlight } from "./desktop-async-control";

const AUTH_ME_PATH = "/api/auth/me";
const ORG_PATH = "/api/org";

type RunAuthWindow = (
  request: DesktopAuthWindowRequest,
) => Promise<string | null>;

interface DesktopAuthSessionOptions {
  /** Pre-resolved API base URL (`resolveComputerUseApiBaseUrl(platformUrl)`). */
  readonly apiBaseUrl: string;
  readonly addClientHeaders: DesktopClientHeaderInjector;
  /** `buildDesktopAuthTokenUrl(authUrl)`. */
  readonly tokenUrl: string;
  /** `buildDesktopAuthConsumeUrl(authUrl, code, handoffId)`. */
  readonly consumeUrl: (code: string, handoffId: string | null) => string;
  /** `buildDesktopAuthSelectOrgUrl(authUrl, true)`. */
  readonly selectOrgUrl: string;
  readonly runAuthWindow: RunAuthWindow;
  /** Zero-arg "something changed" signal; defaults to a no-op. */
  readonly onChange?: () => void;
  /**
   * Invoked after an interactive consume / org-selection flow completes, so the
   * caller can restart dependent runtimes. Background token refresh does NOT
   * trigger it.
   */
  readonly onAuthCompleted?: (signal: AbortSignal) => Promise<void> | void;
}

function signedOutDesktopAuthState(): DesktopAuthState {
  return {
    status: "signed_out",
    user: null,
    organization: null,
  };
}

function signingInDesktopAuthState(): DesktopAuthState {
  return {
    status: "signing_in",
    user: null,
    organization: null,
  };
}

/**
 * Owns the desktop auth token state machine, extracted from `main.ts` and kept
 * free of Electron imports so it can be integration-tested by injecting fakes,
 * mirroring `ComputerUseHostRuntime`'s dependency-injection shape.
 */
export class DesktopAuthSession {
  private readonly apiBaseUrl: string;
  private readonly addClientHeaders: DesktopClientHeaderInjector;
  private readonly tokenUrl: string;
  private readonly consumeUrl: (
    code: string,
    handoffId: string | null,
  ) => string;
  private readonly selectOrgUrl: string;
  private readonly runAuthWindow: RunAuthWindow;
  private readonly onChange: () => void;
  private readonly onAuthCompleted: (
    signal: AbortSignal,
  ) => Promise<void> | void;

  private token: string | null = null;
  private lifetime = new AbortController();
  private appState: DesktopAuthState = signedOutDesktopAuthState();
  private readonly tokenRefresh = singleFlight(() => this.refreshToken());
  private pendingCallback: DesktopAuthCallback | null = null;
  private signingIn = false;
  private restoreEnabled = true;
  private authority: {
    readonly userId: string;
    readonly orgId: string;
    readonly lifetime: AbortSignal;
  } | null = null;

  /** Opaque, current identity/session proof. Never sent over the renderer bridge. */
  getAuthority(): object | null {
    return !this.signingIn && !this.lifetime.signal.aborted
      ? this.authority
      : null;
  }

  private rememberAuthority(
    state: DesktopAuthState,
    lifetime: AbortController,
  ): void {
    if (this.lifetime !== lifetime || lifetime.signal.aborted) return;
    const next =
      state.status === "signed_in" && state.organization
        ? {
            userId: state.user.userId,
            orgId: state.organization.id,
            lifetime: lifetime.signal,
          }
        : null;
    if (
      this.authority?.userId === next?.userId &&
      this.authority?.orgId === next?.orgId &&
      this.authority?.lifetime === next?.lifetime
    )
      return;
    this.authority = next;
    this.onChange();
  }

  constructor(options: DesktopAuthSessionOptions) {
    this.apiBaseUrl = options.apiBaseUrl;
    this.addClientHeaders = options.addClientHeaders;
    this.tokenUrl = options.tokenUrl;
    this.consumeUrl = options.consumeUrl;
    this.selectOrgUrl = options.selectOrgUrl;
    this.runAuthWindow = options.runAuthWindow;
    this.onChange = options.onChange ?? (() => {});
    this.onAuthCompleted = options.onAuthCompleted ?? (() => {});
  }

  async getToken(options?: {
    readonly forceRefresh?: boolean;
  }): Promise<string | null> {
    if (this.signingIn) return null;
    if (!options?.forceRefresh && this.token) {
      return this.token;
    }
    if (!this.restoreEnabled) {
      return null;
    }
    return await this.refresh();
  }

  /**
   * `init` carries the method and body for non-GET calls. Its `headers` are
   * merged under the session headers, which always win, so a caller cannot
   * accidentally drop the bearer token.
   */
  async fetchWithSessionAuth(
    requestUrl: URL,
    init?: RequestInit,
  ): Promise<Response> {
    return await this.fetchWithAppAuth(requestUrl, init);
  }

  getCachedToken(): string | null {
    return this.token;
  }

  async getAuthState(): Promise<DesktopAuthState> {
    if (this.signingIn) {
      return signingInDesktopAuthState();
    }
    if (!this.restoreEnabled) {
      return signedOutDesktopAuthState();
    }

    return await this.getAppAuthState();
  }

  signOut(): void {
    this.lifetime.abort();
    this.authority = null;
    this.token = null;
    this.appState = signedOutDesktopAuthState();
    this.tokenRefresh.clear();
    this.pendingCallback = null;
    this.signingIn = false;
    this.restoreEnabled = false;
    this.onChange();
  }

  async consumeCode(
    code: string,
    handoffId: string | null = null,
  ): Promise<void> {
    this.tokenRefresh.clear();
    await this.authenticate(this.consumeUrl(code, handoffId), true, false);
  }

  async selectOrganization(): Promise<void> {
    this.tokenRefresh.clear();
    await this.authenticate(this.selectOrgUrl, true, true);
  }

  /**
   * Fire-and-forget consume of a parsed auth callback, so callers just hand
   * the session a callback instead of re-implementing the consume plumbing.
   */
  consumeCallback(
    callback: DesktopAuthCallback,
    onError: (error: unknown) => void,
  ): void {
    void this.consumeCode(callback.code, callback.handoffId).catch(onError);
  }

  queuePendingCallback(callback: DesktopAuthCallback): void {
    this.pendingCallback = callback;
  }

  takePendingCallback(): DesktopAuthCallback | null {
    const callback = this.pendingCallback;
    this.pendingCallback = null;
    return callback;
  }

  private async refresh(): Promise<string | null> {
    return await this.tokenRefresh();
  }

  private async refreshToken(): Promise<string | null> {
    try {
      return await this.authenticate(this.tokenUrl, false, false);
    } catch {
      // A failed App restoration requires explicit sign-in, never another
      // identity source. Interactive failures still reject to the caller.
      return null;
    }
  }

  private async authenticate(
    url: string,
    interactive: boolean,
    visible: boolean,
  ): Promise<string | null> {
    this.lifetime.abort();
    const lifetime = new AbortController();
    this.lifetime = lifetime;
    this.authority = null;
    this.restoreEnabled = true;
    this.token = null;
    this.appState = signedOutDesktopAuthState();
    this.setSigningIn(interactive);
    // Even a hidden token restoration replaces session execution authority.
    this.onChange();
    // The lifetime also owns validation requests after the window closes.
    const signal = AbortSignal.any([
      lifetime.signal,
      AbortSignal.timeout(30_000),
    ]);
    try {
      const token = await this.runAuthWindow({
        url,
        visible,
        allowInteractiveFallbacks: interactive,
        signal,
      });
      signal.throwIfAborted();
      if (!token) return null;
      const state = await this.readAppIdentity(token, signal);
      signal.throwIfAborted();
      if (state.status !== "signed_in") return null;
      this.appState = state;
      this.rememberAuthority(state, lifetime);
      this.token = token;
      this.onChange();
      if (interactive) {
        this.setSigningIn(false);
        await this.onAuthCompleted(lifetime.signal);
        lifetime.signal.throwIfAborted();
      }
      return token;
    } catch (error) {
      if (this.lifetime === lifetime) {
        this.token = null;
        this.appState = signedOutDesktopAuthState();
      }
      if (!interactive && signal.aborted) return null;
      throw error;
    } finally {
      if (this.lifetime === lifetime) {
        if (!this.token) {
          // Notify renderer/tray subscribers, and keep their reads from
          // reopening a failed hidden restore until explicit sign-in.
          this.restoreEnabled = false;
          this.onChange();
        }
        this.setSigningIn(false);
      }
    }
  }

  private async appRequest(
    requestUrl: URL,
    token: string,
    signal: AbortSignal,
    init?: RequestInit,
  ): Promise<Response> {
    if (requestUrl.origin !== new URL(this.apiBaseUrl).origin) {
      throw new Error("Invalid Desktop API origin");
    }
    signal.throwIfAborted();
    const headers = new Headers(init?.headers);
    headers.delete("cookie");
    headers.set("authorization", `Bearer ${token}`);
    this.addClientHeaders(headers);
    const response = await fetch(requestUrl, {
      ...init,
      headers,
      credentials: "omit",
      redirect: "error",
      signal: init?.signal ? AbortSignal.any([signal, init.signal]) : signal,
    });
    signal.throwIfAborted();
    return response;
  }

  private async readAppIdentity(
    token: string,
    signal: AbortSignal,
  ): Promise<DesktopAuthState> {
    const me = await this.appRequest(
      new URL(AUTH_ME_PATH, this.apiBaseUrl),
      token,
      signal,
    );
    if (me.status === 401) return signedOutDesktopAuthState();
    if (!me.ok) throw new Error(`Desktop auth status failed: ${me.status}`);
    const user = authContract.me.responses[200].parse(await me.json());
    signal.throwIfAborted();
    if (!user.userId || !user.orgId) return signedOutDesktopAuthState();
    // Both reads use the identical server-verified bearer; never refresh only
    // the second half of the user/workspace pair.
    const org = await this.appRequest(
      new URL(ORG_PATH, this.apiBaseUrl),
      token,
      signal,
    );
    if (org.status === 401 || org.status === 404)
      return signedOutDesktopAuthState();
    if (!org.ok)
      throw new Error(`Desktop organization status failed: ${org.status}`);
    const organization: unknown = await org.json();
    signal.throwIfAborted();
    if (
      typeof organization !== "object" ||
      organization === null ||
      !("id" in organization) ||
      organization.id !== user.orgId ||
      !("name" in organization) ||
      typeof organization.name !== "string"
    ) {
      return signedOutDesktopAuthState();
    }
    return {
      status: "signed_in",
      user: { userId: user.userId, email: user.email },
      organization: { id: user.orgId, name: organization.name },
    };
  }

  private async getAppAuthState(): Promise<DesktopAuthState> {
    if (!this.token) {
      await this.getToken();
      return this.appState;
    }
    const lifetime = this.lifetime;
    try {
      const state = await this.readAppIdentity(this.token, lifetime.signal);
      lifetime.signal.throwIfAborted();
      if (state.status === "signed_in") {
        this.appState = state;
        this.rememberAuthority(state, lifetime);
        return state;
      }
      await this.getToken({ forceRefresh: true });
      return this.appState;
    } catch (error) {
      if (lifetime.signal.aborted) return signedOutDesktopAuthState();
      this.rememberAuthority(signedOutDesktopAuthState(), lifetime);
      throw error;
    }
  }

  private async fetchWithAppAuth(
    requestUrl: URL,
    init?: RequestInit,
  ): Promise<Response> {
    const token = await this.getToken();
    if (!token || token !== this.token)
      return new Response(null, { status: 401 });
    const lifetime = this.lifetime;
    const response = await this.appRequest(
      requestUrl,
      token,
      lifetime.signal,
      init,
    );
    if (response.status !== 401) return response;
    // No cookie-only retry. One App refresh and at most one authenticated retry.
    const refreshed = await this.getToken({ forceRefresh: true });
    if (!refreshed || refreshed !== this.token) return response;
    const retried = await this.appRequest(
      requestUrl,
      refreshed,
      this.lifetime.signal,
      init,
    );
    if (retried.status === 401) {
      this.restoreEnabled = false;
      this.token = null;
      this.appState = signedOutDesktopAuthState();
      this.authority = null;
      this.onChange();
    }
    return retried;
  }

  private setSigningIn(value: boolean): void {
    if (this.signingIn === value) {
      return;
    }
    this.signingIn = value;
    this.onChange();
  }
}

File: turbo/apps/desktop/native/OkouDesktop/Sources/OkouDesktopKit/Auth/DesktopAuthSession.swift
import Foundation
#if canImport(FoundationNetworking)
import FoundationNetworking
#endif

public struct DesktopAuthWindowRequest: Equatable, Sendable {
    public let url: String
    public let visible: Bool
    public let allowInteractiveFallbacks: Bool

    public init(url: String, visible: Bool, allowInteractiveFallbacks: Bool) {
        self.url = url
        self.visible = visible
        self.allowInteractiveFallbacks = allowInteractiveFallbacks
    }
}

public struct DesktopAuthError: Error, Equatable, CustomStringConvertible {
    public let message: String

    public init(_ message: String) {
        self.message = message
    }

    public var description: String { message }
}

/// Owns the desktop auth token state machine. Port of `DesktopAuthSession`:
/// the token is delivered out of band by `completeSignIn` from the hidden
/// WebKit window, lives in memory only, and is re-minted through the token
/// page when a request comes back 401.
@MainActor
public final class DesktopAuthSession {
    public static let authMePath = "/api/auth/me"
    public static let orgPath = "/api/org"

    private let apiBaseUrl: String
    private let cookieUrls: [URL]
    private let cookieSource: DesktopSessionCookieSource
    private let http: DesktopHTTPClient
    private let clientHeaders: DesktopClientHeaders
    private let tokenUrl: String
    private let consumeUrl: (String, String?) -> String
    private let selectOrgUrl: String
    private let runAuthWindow: (DesktopAuthWindowRequest) async throws -> Void
    private let onChange: () -> Void
    private let onAuthCompleted: () async throws -> Void

    private var token: String? = nil
    private var refreshTask: Task<String?, Error>? = nil
    private var pendingCallback: DesktopAuthCallback? = nil
    private var signingIn = false
    private var acceptsSignInCompletions = true

    public init(
        apiBaseUrl: String,
        cookieUrls: [URL],
        cookieSource: DesktopSessionCookieSource,
        http: DesktopHTTPClient,
        clientHeaders: DesktopClientHeaders,
        tokenUrl: String,
        consumeUrl: @escaping (String, String?) -> String,
        selectOrgUrl: String,
        runAuthWindow: @escaping (DesktopAuthWindowRequest) async throws -> Void,
        onChange: @escaping () -> Void = {},
        onAuthCompleted: @escaping () async throws -> Void = {}
    ) {
        self.apiBaseUrl = apiBaseUrl
        self.cookieUrls = cookieUrls
        self.cookieSource = cookieSource
        self.http = http
        self.clientHeaders = clientHeaders
        self.tokenUrl = tokenUrl
        self.consumeUrl = consumeUrl
        self.selectOrgUrl = selectOrgUrl
        self.runAuthWindow = runAuthWindow
        self.onChange = onChange
        self.onAuthCompleted = onAuthCompleted
    }

    public var cachedToken: String? {
        token
    }

    public var isSigningIn: Bool {
        signingIn
    }

    public func getToken(forceRefresh: Bool = false) async throws -> String? {
        if !forceRefresh, let token {
            return token
        }
        if !acceptsSignInCompletions {
            return nil
        }
        return try await refresh()
    }

    /// Cookies plus cached bearer; on 401 retries with cookies only, then
    /// once more with a freshly minted token.
    public func fetchWithSessionAuth(
        _ requestUrl: URL, method: String = "GET", headers: [String: String] = [:], body: Data? = nil
    ) async throws -> DesktopHTTPResponse {
        let response = try await http.send(await request(requestUrl, method: method, headers: headers, body: body))
        if response.status != 401 || token == nil {
            return response
        }
        token = nil
        let withCookies = try await http.send(await request(requestUrl, method: method, headers: headers, body: body))
        if withCookies.status != 401 {
            return withCookies
        }
        guard try await getToken(forceRefresh: true) != nil else {
            return withCookies
        }
        return try await http.send(await request(requestUrl, method: method, headers: headers, body: body))
    }

    public func getAuthState() async throws -> DesktopAuthState {
        if signingIn {
            return .signingIn
        }
        if !acceptsSignInCompletions {
            return .signedOut
        }
        let hadToken = token != nil
        let state = try await fetchAuthState()
        if state != .signedOut || hadToken {
            return state
        }
        guard try await getToken(forceRefresh: true) != nil else {
            return state
        }
        return try await fetchAuthState()
    }

    private func fetchAuthState() async throws -> DesktopAuthState {
        let meResponse = try await fetchWithSessionAuth(URL(string: apiBaseUrl + Self.authMePath)!)
        if meResponse.status == 401 {
            return .signedOut
        }
        guard meResponse.ok else {
            throw DesktopAuthError("Desktop auth status failed: \(meResponse.status)")
        }
        let me = try meResponse.json()
        let user = DesktopAuthUser(userId: me["userId"]?.stringValue ?? "", email: me["email"]?.stringValue ?? "")
        let orgResponse = try await fetchWithSessionAuth(URL(string: apiBaseUrl + Self.orgPath)!)
        if orgResponse.status == 401 {
            return .signedOut
        }
        if orgResponse.status == 404 {
            return .signedIn(user: user, organization: nil)
        }
        guard orgResponse.ok else {
            throw DesktopAuthError("Desktop organization status failed: \(orgResponse.status)")
        }
        let org = try orgResponse.json()
        return .signedIn(
            user: user,
            organization: DesktopAuthOrganization(id: org["id"]?.stringValue ?? "", name: org["name"]?.stringValue ?? "")
        )
    }

    /// The token's only write path; ignored after sign-out until the next
    /// interactive flow re-arms the session.
    public func completeSignIn(token: String) {
        guard acceptsSignInCompletions else { return }
        self.token = token
        onChange()
    }

    public func signOut() {
        token = nil
        refreshTask = nil
        pendingCallback = nil
        signingIn = false
        acceptsSignInCompletions = false
        onChange()
    }

    public func consumeCode(_ code: String, handoffId: String? = nil) async throws {
        acceptsSignInCompletions = true
        setSigningIn(true)
        do {
            try await runAuthWindow(DesktopAuthWindowRequest(url: consumeUrl(code, handoffId), visible: false, allowInteractiveFallbacks: true))
        } catch {
            setSigningIn(false)
            throw error
        }
        setSigningIn(false)
        try await onAuthCompleted()
    }

    public func selectOrganization() async throws {
        acceptsSignInCompletions = true
        try await runAuthWindow(DesktopAuthWindowRequest(url: selectOrgUrl, visible: true, allowInteractiveFallbacks: true))
        try await onAuthCompleted()
    }

    public func consumeCallback(_ callback: DesktopAuthCallback, onError: @escaping (Error) -> Void) {
        Task { @MainActor in
            do {
                try await self.consumeCode(callback.code, handoffId: callback.handoffId)
            } catch {
                onError(error)
            }
        }
    }

    public func queuePendingCallback(_ callback: DesktopAuthCallback) {
        pendingCallback = callback
    }

    public func takePendingCallback() -> DesktopAuthCallback? {
        let callback = pendingCallback
        pendingCallback = nil
        return callback
    }

    private func refresh() async throws -> String? {
        if let refreshTask {
            return try await refreshTask.value
        }
        let task = Task<String?, Error> { @MainActor in
            try await self.refreshToken()
        }
        refreshTask = task
        defer {
            if refreshTask == task {
                refreshTask = nil
            }
        }
        return try await task.value
    }

    private func refreshToken() async throws -> String? {
        let before = token
        try await runAuthWindow(DesktopAuthWindowRequest(url: tokenUrl, visible: false, allowInteractiveFallbacks: false))
        let after = token
        // A refresh window that completed without delivering a token leaves
        // the token unchanged; report nil rather than resending a stale one.
        return after == before ? nil : after
    }

    private func setSigningIn(_ value: Bool) {
        if signingIn == value { return }
        signingIn = value
        onChange()
    }

    private func request(_ url: URL, method: String, headers: [String: String], body: Data?) async -> URLRequest {
        var request = URLRequest.desktop(url: url, method: method, headers: headers, body: body)
        let cookieHeader = await DesktopSessionCookies.cookieHeader(source: cookieSource, urls: cookieUrls + [url])
        request.applyDesktopSessionHeaders(cookieHeader: cookieHeader, token: token, clientHeaders: clientHeaders)
        return request
    }
}

/// Session-authenticated fetch for the host runtime's `hosts/start` and
/// `/api/auth/me` calls: cookies for the platform URL and the request URL,
/// the cached bearer, and one forced refresh on 401. Port of
/// `createDesktopComputerUseSessionFetch`.
public struct ComputerUseSessionFetch: Sendable {
    public let platformUrl: URL
    public let cookieSource: DesktopSessionCookieSource
    public let http: DesktopHTTPClient
    public let clientHeaders: DesktopClientHeaders
    public let getCachedAuthToken: @Sendable () async -> String?
    public let getAuthToken: @Sendable (Bool) async throws -> String?

    public init(
        platformUrl: URL, cookieSource: DesktopSessionCookieSource, http: DesktopHTTPClient,
        clientHeaders: DesktopClientHeaders,
        getCachedAuthToken: @escaping @Sendable () async -> String?,
        getAuthToken: @escaping @Sendable (Bool) async throws -> String?
    ) {
        self.platformUrl = platformUrl
        self.cookieSource = cookieSource
        self.http = http
        self.clientHeaders = clientHeaders
        self.getCachedAuthToken = getCachedAuthToken
        self.getAuthToken = getAuthToken
    }

    public func send(_ request: URLRequest) async throws -> DesktopHTTPResponse {
        guard let url = request.url else {
            throw DesktopAuthError("Session request is missing a URL")
        }
        let cookieHeader = await DesktopSessionCookies.cookieHeader(source: cookieSource, urls: [platformUrl, url])
        var first = request
        first.applyDesktopSessionHeaders(cookieHeader: cookieHeader, token: await getCachedAuthToken(), clientHeaders: clientHeaders)
        let response = try await http.send(first)
        if response.status != 401 {
            return response
        }
        guard let refreshed = try await getAuthToken(true) else {
            return response
        }
        var retry = request
        retry.applyDesktopSessionHeaders(cookieHeader: cookieHeader, token: refreshed, clientHeaders: clientHeaders)
        return try await http.send(retry)
    }

    public var fetch: DesktopFetch {
        { request in try await send(request) }
    }
}
